src/ucs/config/parser.c

```diff
@@ -1066,6 +1066,173 @@ void ucs_config_help_table(char *buf, size_t max, const void *arg)
     snprintf(buf, max, "Table");
 }
 
+/**
+ * Return number of allowed keys in the given config
+ */
+static size_t ucs_config_key_value_count(const ucs_config_key_field_t *cfg_key)
+{
+    size_t count = 0;
+    for (; cfg_key->name; ++cfg_key, ++count);
+    return count;
+}
+
+/**
+ * Given key-value pair, find appropriate destination for storing value in the
+ * temporary values buffer, and save it there
+ */
+static ucs_status_t
+ucs_config_key_value_set(const ucs_config_key_field_t *cfg_key,
+                         const char **values, size_t key_count,
+                         const char *key, const char *value)
+{
+    size_t index;
+
+    if (NULL == value) {
+        /* Empty value means that key is a generic value for all keys */
+        for (index = 0; index < key_count; ++index) {
+            values[index] = key;
+        }
+        return UCS_OK;
+    } else {
+        /* Key configured, save value for specific key */
+        for (index = 0; index < key_count; ++index, ++cfg_key) {
+            if (!strcmp(cfg_key->name, key)) {
+                values[index] = value;
+                return UCS_OK;
+            }
+        }
+        return UCS_ERR_INVALID_PARAM;
+    }
+}
+
+int ucs_config_sscanf_key_value(const char *buf, void *dest, const void *arg)
+{
+    const ucs_config_key_value_param_t *param = arg;
+    size_t buf_len                            = strlen(buf) + 1;
+    const char *TOKEN_DELIM                   = ",";
+    const char *VALUE_DELIM                   = ":";
+    size_t key_count, index;
+    const char **values;
+    char *token, *tokens;
+    const char *key, *value;
+
+    /* values is a temporary array value storage for the keys.
+     * In values array we remember the last applicable setting for each key.
+     * We need it to perform actual setting exactly once per each key */
+    key_count = ucs_config_key_value_count(param->keys);
+    values = ucs_alloca(key_count * sizeof(*values));
+    memset(values, 0, key_count * sizeof(*values));
```

Comment:
i think we can make it simpler:
- go over the tokens in the input string
- for each key:value pair:
- if key is provided - find key index in the list and save value in "dest" in the right place by calling param->parser.read, and mark it as initialized (in a bitmap array / uint8_t array on the stack)
- otherwise, if key is not provided (value only) - read the value to a temp buffer (the default value)
- after going over the tokens, assign the default value to each un-initialized entry in "dest"

Reply:
Ok, initially this looks good to me.
QQ about bitmaps: do we have an easy way to allocate bitmap of arbitrary size on stack?
Looking at bitmap.h I didn't find such API, but probably just overlooked.

In past I've created this super-simple bitset, that allows to create any dimensional bitmaps on stack:
```
#define TL3_BIT_MASK(bit)       (1 << ((bit) & (CHAR_BIT - 1)))
#define TL3_BIT_SET(set, bit)   (set[(bit)/CHAR_BIT] |= TL3_BIT_MASK(bit))
#define TL3_BIT_RESET(set, bit) (set[(bit)/CHAR_BIT] &= ~TL3_BIT_MASK(bit))
#define TL3_BIT_GET(set, bit)   (set[(bit)/CHAR_BIT] &  TL3_BIT_MASK(bit))
#define TL3_BITSET_SIZE(bits)   ((bits + CHAR_BIT - 1) / CHAR_BIT)

    uint8_t static_bitset[TL3_BITSET_SIZE(compile_time_size)];
    uint8_t *dynamic_bitset = alloca(TL3_BITSET_SIZE(dynamic_size));
```
would it be useful to have in UCX?

Reply:
actually I plan to add dynamic bitmap based on ucs array